resource_detectors/lib/opentelemetry/resource/detectors/container.rb

```diff
@@ -0,0 +1,59 @@
+# frozen_string_literal: true
+
+# Copyright The OpenTelemetry Authors
+#
+# SPDX-License-Identifier: Apache-2.0
+
+module OpenTelemetry
+  module Resource
+    module Detectors
+      # Container contains detect class method for determining container resource attributes
+      module Container
+        extend self
+
+        UUID_PATTERN = '[0-9a-f]{8}[-_]?[0-9a-f]{4}[-_]?[0-9a-f]{4}[-_]?[0-9a-f]{4}[-_]?[0-9a-f]{12}'
+        CONTAINER_PATTERN = '[0-9a-f]{64}'
+        CONTAINER_REGEX = /(?<container>#{UUID_PATTERN}|#{CONTAINER_PATTERN})(?:.scope)?$/
+        CGROUP_V1_PATH = '/proc/self/cgroup'
+        CGROUP_V2_PATH = '/proc/self/mountinfo'
+
+        # Detects container attributes and creates a {OpenTelemetry::SDK::Resources::Resource} with the specified attributes.
+        # If no container attributes could be determined an empty resource is returned
+        #
+        # @return [OpenTelemetry::SDK::Resources::Resource]
+        def detect
+          id = container_id
+          resource_attributes = {}
+
+          resource_attributes[OpenTelemetry::SemanticConventions::Resource::CONTAINER_ID] = id unless id.nil?
+          resource_attributes.delete_if { |_key, value| value.nil? || value.empty? }
+          OpenTelemetry::SDK::Resources::Resource.create(resource_attributes)
+        end
+
+        private
+
+        # Returns the container.id if it can be determined from cgroup
+        # or nil if container.id could not be determined
+        #
+        # @return [String] container.id
+        #   May be nil.
+        def container_id
+          [CGROUP_V2_PATH, CGROUP_V1_PATH].each do |cgroup|
+            next unless File.readable?(cgroup)
+
+            File.readlines(cgroup, chomp: true).each do |line|
+              parts = line.split('/')
```

Comment:
Is there some nuance I am missing here. Does scanning the full string result in false positives?

Reply:
scanning the full string works for cgroupv1; but I couldn't get it to work for cgroupv2. This way it's compatible with both